Bitsets3 (#239)
fortran-lang/stdlib: doc/specs/stdlib_bitsets.md

`int32`, and `int64`. In the assignment to and from logical arrays
array index, `i`, is mapped to bit position, `pos=i-1`, and `.true.`
is mapped to a set bit, and `.false.` is mapped to an unset bit.

#### Example

    ```fortran
    program demo_assignment
        use stdlib_bitsets
        logical(int8)  :: logical1(64) = .true.
        logical(int32), allocatable :: logical2(:)
        type(bitset_64) :: set0, set1
        set0 = logical1
        if ( set0 % bits() /= 64 ) then
            error stop procedure // &
                ' initialization with logical(int8) failed to set' // &
                ' the right size.'
        else if ( .not. set0 % all() ) then
            error stop procedure // ' initialization with' // &
                ' logical(int8) failed to set the right values.'
        else
            write(*,*) 'Initialization with logical(int8) succeeded.'
        end if
        set1 = set0
        if ( set1 == set0 ) &
            write(*,*) 'Initialization by assignment succeeded'
        logical2 = set1
        if ( all( logical2 ) ) then
            write(*,*) 'Initialization of logical(int32) succeeded.'
        end if
    end program demo_assignment

> 

Reviewer: the ``` are missing
```suggestion
```
```

Subroutine

####  Arguments

`new`: shall be a scalar `bitset_64` or  `bitset_large` variable. It
is an `intent(out)` argument. It will be the new bitset.

`old`: shall be a scalar expression of the same type as `new`. It is
an `intent(in)` argument. It will be the source bitset.

`start_pos`: shall be a scalar integer expression of the kind
`bits_kind`. It is an `intent(in)` argument.

`stop_pos`: shall be a scalar integer expression of the kind
`bits_kind`. It is an `intent(in)` argument.

`status` (optional): shall be a scalar default integer variable. It is
an `intent(out)` argument. If present it shall have one of the values:

* `success` - no problems found

* `index_invalid_error` - `start_pos` was less than zero or `stop_pos`
  was greater than `bits(old)-1`.

#### Example

    ```fortran
    program demo_extract
        use stdlib_bitsets
        type(bitset_large) :: set0, set1
        call set0 % init(166)
        call set0 % set(100,150)
        call extract( set1, set0, 100, 150)
        if ( set1 % bits() == 51 ) &
            write(*,*) 'SET1 has the proper size.'
        if ( set1 % all() ) write(*,*) 'SET1 has the proper values.'
    end program demo_extract

> 

Reviewer: The ``` are missing

flip the bit values with positions from `start_pos` to `end_pos` in
`self`.
* If `end_pos < start_pos` then `self` is unmodified.

#### Syntax

`call self % [[bitset_type(class):flip(bound)]] (pos)`

or

`call self % [[bitset_type(class):flip(bound)]] (start_pos, end_pos)`

#### Class

Elemental subroutine.

#### Arguments

`self`: shall be a scalar class `bitset_type` variable It is an
`intent(inout)` argument.

`pos`: shall be a scalar integer expression of kind `bits_kind`. It is
an `intent(in)` argument.

`start_pos`: shall be a scalar integer expression of kind
`bits_kind`. It is an `intent(in)` argument.

`end_pos`: shall be a scalar integer expression of kind
`bits_kind`. It is an `intent(in)` argument.

#### Example

    ```fortran
    program demo_flip
        use stdlib_bitsets
        type(bitset_large) :: set0
        call set0 % init(166)
        if ( set0 % none() ) write(*,*) 'SET0 is properly initialized.'
        call set0 % flip(165)
        if ( set0 % test(165) ) write(*,*) 'Bit 165 is flipped.'
        call set0 % flip(0,164)
        if ( set0 % all() ) write(*,*) 'All bits are flipped.'
    end program demo_flip

> 

Reviewer: the ``` are missing

        if ( all(set0) ) write(*,*) 'Third test of OR worked.'
        call set0 % not()
        call or( set0, set1 ) ! all all
        if ( all(set0) ) write(*,*) 'Fourth test of OR worked.'
    end program demo_or
```

### `output` - Writes a binary representation of a bitset to a file

#### Status

Experimental

#### Description

Writes a binary representation of a bitset to an unformatted file.

#### Syntax

`call self % [[bitset_type(class):output(bound)]] (unit[, status])`

#### Class

Subroutine.

#### Arguments

`self`: shall be a scalar expression of class `bitset_64` or
`bitset_large`. It is an `intent(in)` argument.

`unit`: shall be a scalar default integer expression. It is an
`intent(in)` argument. Its value must be that of an I/O unit number
for an open unformatted file with `WRITE` or `READWRITE` access.

`status` (optional): shall be a scalar default integer variable. It is
an `intent(out)` argument. If present on return it will have the value
of `success` or `write_failure`. If absent and it would not have the
value of `success` then processing will stop with an informative stop
code. The two code values have the meaning:

* `success` - no problem found

> * `write_failure` - a failure occured in a write statement.

Reviewer: ```suggestion
* `write_failure` - a failure occurred in a write statement.
```

---
title: Bitsets
---

# The `stdlib_bitsets` module

[TOC]

## Introduction

The `stdlib_bitsets` module implements bitset types. A bitset is a
compact representation of a sequence of `bits` binary values. It can
equivalently be considered as a sequence of logical values or as a

> subset of the integers 0 ... `bits-1`. The bits are indexed from 0 to

Reviewer 2: > or as a subset of the integers 0 ... `bits-1`.

This confuses me. The equivalent integer values are 0 and 1, not 0 through `bits-1`, correct?

Do you mean to say something like this?

"It can equivalently be considered as a sequence of logical values or as a sequence of integers 0 and 1 with indices in the range 0... `bits-1`?"

|`integer_overflow_error`|Attempt to define an integer value bigger than `huge(0_bits_kind`)|
|`read_failure`|Failure on a `read` statement|
|`eof_failure`|An unexpected "End-of-File" on a `read` statement|
|`write_failure`|Failure on a `write` statement|

## The `stdlib_bitsets` derived types

The `stdlib_bitsets` module defines three derived types,
`bitset_type`, `bitset_64`, and `bitset_large`. `bitset_type` is an abstract
type that serves as the ancestor of `bitset_64` and
`bitset_large`. `bitset_type` defines one method, `bits`, all of its
other methods are deferred to its extensions. `bitset_64` is a bitset
that can handle up 64 bits. `bitset_large` is a bitset that can handle
up `huge(0_bits_kind)` bits. All attributes of the bitset types are
private. The various types each define a sequence of binary values: 0
or 1. In some cases it is useful to associate a logical value, `test`,
for each element of the sequence, where `test` is `.true.` if the value
is 1 and `.false.` otherwise. The number of such values in an entity
of that type is to be termed, `bits`. The bits are ordered in terms of
position, that, in turn, is indexed from 0 to `bits-1`. `bitset_type` is

> not used in source code. The syntax for using the types are:

Reviewer: ```suggestion
not used in source code of `stdlib`, but is publicaly avaliable fo users to represent either `bitset_64` or `bitset_large`.
The syntax for using the types are:
```
Related to [this comment](https://github.com/fortran-lang/stdlib/pull/239/files#r504891872)

Author: I don't think this is correct. I have changed the text to

used only as a `class` to define entities that can be either a `bitset_64` or
a `bitset_large`. The syntax for using the types are:

Committed and pushed.

---
title: Bitsets
---

# The `stdlib_bitsets` module

[TOC]

## Introduction

The `stdlib_bitsets` module implements bitset types. A bitset is a
compact representation of a sequence of  binary values, that can
equivalently be considered a sequence of logical values or a subset of
the integers 0 ... `bits(bitset)-1`. The bits are indexed from 0 to
`bits(bitset)-1`. A bitset is used when space savings are critical in applications that require a large number
of closely related logical values.
It may also improve performance by reducing memory traffic. To implement bitsets the module
defines three bitset types, multiple constants, a character string
literal that can be read to and from strings and formatted files, a
simple character string literal that can be read to and from strings,
assignments, procedures, methods, and operators. Note that the module

> assumes two's complement integers, but all current Fortran 95+ processors use such integers.

Author: It is necessary if users are concerned about the portability of their code. There are still some Unisys machines in operation, and they used one's complement, but never had a Fortran 90+ compiler. There may also be some microcontrollers with one's complement or signed magnitude implementations.

> assumes two's complement integers, but all current Fortran 95+ processors use such integers.

Reviewer: Thank you. I 'll remove my suggestion.

## The module's constants

The module defines several public constants all integers, almost all
intended to serve as error codes in reporting problems through an
optional `stat` argument. One constant, `bits_kind` is
the integer kind value for indexing bits and reporting counts of
bits. The other constants that are error codes are summarized below:

> |Error Code|Summary|

Author: Wait until the reviewers have seen what I did in the source code.

|`integer_overflow_error`|Attempt to define an integer value bigger than `huge(0_bits_kind`)|
|`read_failure`|Failure on a `read` statement|
|`eof_failure`|An unexpected "End-of-File" on a `read` statement|
|`write_failure`|Failure on a `write` statement|

## The `stdlib_bitsets` derived types

The `stdlib_bitsets` module defines three derived types,
`bitset_type`, `bitset_64`, and `bitset_large`. `bitset_type` is an abstract
type that serves as the ancestor of `bitset_64` and
`bitset_large`. `bitset_type` defines one method, `bits`, all of its
other methods are deferred to its extensions. `bitset_64` is a bitset
that can handle up 64 bits. `bitset_large` is a bitset that can handle
up `huge(0_bits_kind)` bits. All attributes of the bitset types are
private. The various types each define a sequence of binary values: 0
or 1. In some cases it is useful to associate a logical value, `test`,
for each element of the sequence, where `test` is `.true.` if the value
is 1 and `.false.` otherwise. The number of such values in an entity
of that type is to be termed, `bits`. The bits are ordered in terms of

> position, that, in turn, is indexed from 0 to `bits-1`. `bitset_type` is

Author: It can be used to declare a class variable (`class(bitset_type)`) that can be used to represent either bitset_64 or bitset_large types.

`int32`, and `int64`. In the assignment to and from logical arrays
array index, `i`, is mapped to bit position, `pos=i-1`, and `.true.`
is mapped to a set bit, and `.false.` is mapped to an unset bit.

#### Example

```fortran
    program demo_assignment
        use stdlib_bitsets
        logical(int8)  :: logical1(64) = .true.
        logical(int32), allocatable :: logical2(:)
        type(bitset_64) :: set0, set1
        set0 = logical1
        if ( set0 % bits() /= 64 ) then
            error stop procedure // &
                ' initialization with logical(int8) failed to set' // &
                ' the right size.'
        else if ( .not. set0 % all() ) then
            error stop procedure // ' initialization with' // &
                ' logical(int8) failed to set the right values.'
        else
            write(*,*) 'Initialization with logical(int8) succeeded.'
        end if
        set1 = set0
        if ( set1 == set0 ) &
            write(*,*) 'Initialization by assignment succeeded'
        logical2 = set1
        if ( all( logical2 ) ) then
            write(*,*) 'Initialization of logical(int32) succeeded.'
        end if
    end program demo_assignment
```

> ### Table of the `bitset_type` methods

Author: OK

`and`, `and_not`, `or`, and `xor`  also require that the two bitset
arguments have the same number of bits, otherwise the results are
undefined, These procedures are summarized in the following table:

|Procedure name|Class|Summary|
|--------------|-----|-------|
|`and`|elemental subroutine|Sets `self` to the bitwise `and` of the original bits in `self` and `set2`|
|`and_not`|elemental subroutine|Sets `self` to the bitwise `and` of the original bits in `self` and the negation of `set2`|
|`extract`|subroutine|creates a new bitset, `new`, from a range in `old`|
|`or`|elemental subroutine|Sets `self` to the bitwise `or` of the original bits in `self` and `set2`|
|`xor`|elemental subroutine|Sets `self` to the bitwise exclusive `or` of the original bits in `self` and `set2`|

### Table of the non-member comparison operations
The comparison operators with two arguments of type `bitset_large` or
`bitset_64` must have both arguments of the same known type which
prevents them from being methods. The operands must also have the same
number of bits otherwise the results are undefined. These operators
are summarized in the following table:

> |Operator|Description|

Author: My understanding is that ==, etc. are synonyms for .EQ., etc. and defining one automatically defines the other. If I remember correctly gfortran objected when I tried to define both pairs separately on another project. I could list them in this table for thoroughness.

> |Operator|Description|

Reviewer: Ok. If defining one automatically defines the other, then I think it would be good to add the other ones in the table too.

|`==`|`.true.` if all bits in `set1` and `set2` have the same value, `.false.` otherwise|
|`/=`|`.true.` if any bits in `set1` and `set2` differ in value, `.false.` otherwise|
|`>`|`.true.` if the bits in `set1` and `set2` differ in value and the highest order differing bit is 1 in `set1` and 0 in `set2`, `.false.` otherwise|
|`>=`|`.true.` if the bits in `set1` and `set2` are the same or the highest order differing bit is 1 in `set1` and 0 in `set2`, `.false.` otherwise|
|`<`|`.true.` if the bits in `set1` and `set2` differ in value and the highest order differing bit is 0 in `set1` and 1 in `set2`, `.false.` otherwise|
|`<=`|`.true.` if the bits in `set1` and `set2` are the same or the highest order differing bit is 0 in `set1` and 1 in `set2`, `.false.` otherwise|

## Specification of the `stdlib_bitsets` methods and procedures

### `all` - determine whether all bits are set in `self`.

#### Status

Experimental

#### Description

Determines whether all bits are set to 1 in self.

#### Syntax

`result = self % [[bitset_type(class):all(bound)]]()`

#### Class

Elemental function.

#### Argument

`self`: shall be a scalar expression of class `bitset_type`. It is an

> `intent(in)` argument.

Author: The form I use parallels that used in the standard. Your form would be awkward for optional arguments.

> `intent(in)` argument.

Reviewer: Ok. I 'll remove my comment. Thank you for the explanation.

#### Result character

The result is a default logical scalar.

#### Result value

The result is `.true.` if all bits in `self` are set,
otherwise it is `.false.`.

#### Example

```fortran
    program demo_all
        use stdlib_bitsets
        character(*), parameter :: &
            bits_all = '111111111111111111111111111111111'
        type(bitset_64) :: set0
        call set0 % from_string( bits_all )
        if ( bits(set0) /= 33 ) then
            error stop "FROM_STRING failed to interpret " // &
                'BITS_ALL's size properly."
        else if ( .not. set0 % all() ) then
            error stop "FROM_STRING failed to interpret" // &
                "BITS_ALL's value properly."
        else
            write(*,*) "FROM_STRING transferred BITS_ALL properly" // &
                " into set0."
        end if
    end program demo_all
```

#### `and` - bitwise `and` of the bits of two bitsets.

#### Status

Experimental

#### Description

Sets the bits in `set1` to the bitwise `and` of the original bits in
`set1` and `set2`. Note that `set1` and `set2` must have the same

> number of bits, otherwise the result is undefined.

Author: NaN is solely for floating point (real) numbers. The value of the individual bits in the result is undefined. In practice if `set1` has more bytes than `set2`, then you can have a segmentation fault. If `set2` has more bits than `set1` then bits in `set1` beyond `bits(set1)-1` may be modified, which can have an effect if `set1` is used as the second operand in a binary operation. If `set1` has more bits than `set2`, but the same number of bytes, then the bits beyond `bits(set2)-1` will be used in the operation.

> number of bits, otherwise the result is undefined.

Reviewer: thank you for the explanation. IMO it would be good to add these explanations to the specs. Let see what others think about it.

        call set1 % not()
        call and( set0, set1 ) ! none all
        if ( none(set0) ) write(*,*) 'Third test of AND worked.'
        call set0 % not()
        call and( set0, set1 ) ! all all
        if ( all(set0) ) write(*,*) 'Fourth test of AND worked.'
    end program demo_and
```

### `and_not` - Bitwise `and` of one bitset with the negation of another

#### Status

Experimental

#### Description

Sets the bits of `set1` to bitwise `and` of the bits of `set1` with
the bitwise negation of the corresponding bits of `set2`. Note that
`set1` and `set2` must have the same number of bits, otherwise the
result is undefined.

#### Syntax

`call [[stdlib_bitsets(module):and_not(interface)]](set1, set2)`

#### Class

Elemental subroutine.

#### Arguments

`set1`: shall be a scalar `bitset_64` or `bitset_large` variable. It
is an `intent(inout)` argument. On return the values of the bits in
`set1` are the bitwise `and` of the original bits in `set1` with the
corresponding negation of the bits in `set2`.

`set2`: shall be a scalar expression of the same type as `set1`. It is
an `intent(in)` argument. Note that it should also have the same

> number of bits as `set1` otherwise the result is undefined.

Reviewer: ```suggestion
number of bits as `set1`, otherwise the result is undefined.
```

#### Example

```fortran
    program demo_and_not
        use stdlib_bitsets
        type(bitset_large) :: set0, set1
        call set0 % init(166)
        call set1 % init(166)
        call and_not( set0, set1 ) ! none none
        if ( none(set0) ) write(*,*) 'First test of AND_NOT worked.'
        call set0 % not()
        call and_not( set0, set1 ) ! all none
        if ( all(set0) ) write(*,*) 'Second test of AND_NOT worked.'
        call set0 % not()
        call set1 % not()
        call and_not( set0, set1 ) ! none all
        if ( none(set0) ) write(*,*) 'Third test of AND_NOT worked.'
        call set0 % not()
        call and_not( set0, set1 ) ! all all
        if ( none(set0) ) write(*,*) 'Fourth test of AND_NOT worked.'
    end program demo_and_not
```

### `any` - determine whether any bits are set

#### Status

Experimental

#### Description

Determines whether any bits are set in `self`.

#### Syntax

`result = self % [[bitset_type(class):any(bound)]]()`

#### Class

Elemental function.

#### Argument

`self`: shall be a scalar expression of class `bitset_type`. It is an
`intent(in)` argument.

> #### Result character
> 
> The result is a default logical scalar.
> 
> #### Result value
> 
> The result is `.true.` if any bits in `self` are set, otherwise it
> is `.false.`.

Reviewer: These 2 sections could be merged. Currently it is confusing.
```suggestion
#### Result value

The result is a default logical scalar. The result is `.true.` if any bits in `self` are set, otherwise it
is `.false.`.
```

Author: OK

#### Example

```fortran
    program demo_any
        use stdlib_bitsets
        character(*), parameter :: &
            bits_0 = '0000000000000000000'
        type(bitset_64) :: set0
        call set0 % from_string( bits_0 )
        if ( .not. set0 % any() ) then
            write(*,*) "FROM_STRING interpreted " // &
                "BITS_0's value properly."
        end if
        call set0 % set(5)
        if ( set0 % any() ) then
            write(*,*) "ANY interpreted SET0's value properly."
        end if
    end program demo_any
```

### `bit_count` - return the number of bits that are set

#### Status

Experimental

#### Description

Returns the number of bits that are set to one in `self`.

#### Syntax

`result = self % [[bitset_type(class):bit_count(bound)]] ()`

#### Class

Elemental function.

#### Argument

`self`: shall be a scalar expression of class `bitset_type`. It is an
`intent(in)` argument.

> #### Result character
> 
> The result is an integer scalar of kind `bits_kind`.
> 
> #### Result value
> 
> The result is the number of bits that are set in `self`.
> 

Reviewer: ```suggestion
#### Result value

The result is an integer scalar of kind `bits_kind`, equal to the number of bits that are set in `self`.

```

#### Syntax

`result = self % [[bitset_type(class):bits(bound)]] ()`

#### Class

Elemental function.

#### Argument

`self`: shall be a scalar expression of class `bitset_type`. It is an
`intent(in)` argument.

#### Result character

The result is an integer scalar of kind `bits_kind`.

#### Result value

The result is the number of defined bits in `self`.

#### Example

```fortran
    program demo_bits
        use stdlib_bitsets
        character(*), parameter :: &
            bits_0 = '0000000000000000000'
        type(bitset_64) :: set0
        call set0 % from_string( bits_0 )
        if ( set0 % bits() == 19 ) then
            write(*,*) "FROM_STRING interpreted " // &
                "BITS_0's size properly."
        end if
    end program demo_bits
```

### `clear` - clears a sequence of one or more bits.

#### Status

Experimental

#### Description

* If only `pos` is present, clears the bit with position `pos` in
`self`.

* If `start_pos` and `end_pos` are present with `end_pos >= start_pos`
clears the bits with positions from `start_pos` to `end_pos` in `self`.

* if `start_pos` and `end_pos` are present with `end_pos < start_pos`
`self` is unmodified.

> Note: Positions outside the range 0 to `BITS(SET) -1` are ignored.

Reviewer: What happens with the range  -9 to 'bits(set)+10`?
```suggestion
Note: Positions outside the range 0 to `bits(set) -1` are ignored.
```

Reviewer: Reminder for myself: Are there some tests with a range outside 0-`bits(set)-`?

`end_pos`: shall be a scalar integer expression of kind
`bits_kind`. It is an `intent(in)` argument.

#### Example

```fortran
    program demo_clear
        use stdlib_bitsets
        type(bitset_large) :: set0
        call set0 % init(166)
        call set0 % not()
        if ( set0 % all() ) write(*,*) 'SET0 is properly initialized.'
        call set0 % clear(165)
        if ( .not. set0 % test(165) ) write(*,*) 'Bit 165 is cleared.'
        call set0 % clear(0,164)
        if ( set0 % none() ) write(*,*) 'All bits are cleared.'
    end program demo_clear
```

### `extract` - create a new bitset from a range in an old bitset

#### Status

Experimental

#### Description

Creates a new bitset, `new`, from a range, `start_pos` to `stop_pos`,
in bitset `old`. If `start_pos` is greater than `stop_pos` the new
bitset is empty. If `start_pos` is less than zero or `stop_pos` is
greater than `bits(old)-1` then if `status` is present it has the
value `index_invalid_error`, otherwise processing stops with an
informative message.

#### Syntax

`call [[stdlib_bitsets(module):extract(interface)]](new, old, start_pos, stop_pos, status )`

#### Class

> Subroutine

Reviewer: Why is this a subroutine instead a function, e.g.:
`new = extract(old, start_pos,...)`?

Note: I am fine (and even prefer it) as a subroutine. It is mainly to document the PR.

Author: The use of a subroutine:
1. Makes it easier to enforce that `new` must be the same type as `old`
2. Allows me to return a status flag, which is useful now, and necessary if I ever want to allow `new` and `old` to be mixed types

* `alloc_fault` - if allocation of the bitset failed

* `char_string_too_large_error` - if `string` was too large, or 

* `char_string_invalid_error` - if string had an invalid character.

#### Example

```fortran
    program demo_from_string
        use stdlib_bitsets
        character(*), parameter :: &
            bits_all = '111111111111111111111111111111111'
        type(bitset_64) :: set0
        call set0 % from_string( bits_all )
        if ( bits(set0) /= 33 ) then
            error stop "FROM_STRING failed to interpret " // &
                'BITS_ALL's size properly."
        else if ( .not. set0 % all() ) then
            error stop "FROM_STRING failed to interpret" // &
                "BITS_ALL's value properly."
        else
            write(*,*) "FROM_STRING transferred BITS_ALL properly" // &
                " into set0."
        end if
    end program demo_from_string
```

### `init` - `bitset_type` initialization routines.

#### Status

Experimental

#### Description

`bitset_type` initialization routine.

#### Syntax

`call [[stdlib_bitsets(module):init(interface)]] (self, bits [, status])`

#### Class

Subroutine.

#### Arguments

> `self`: shall be a scalar `bitset_64` or `bitset_large` variable. It

Reviewer: to what are set the bits? All `0`? All `1`? Uninitialized?

is an `intent(out)` argument.

`bits` (optional): shall be a scalar integer expression of kind
`bits_kind`. It is an `intent(in)` argument that if present

> specifies the number of bits in `set`. A negative value or a value
> greater than 64 if `self` is of type `bitset_64` is an error.

Reviewer: ```suggestion
specifies the number of bits in `set`. A negative value, or a value
greater than 64 if `self` is of type `bitset_64`, is an error.
```

`status` (optional): shall be a scalar default integer variable. It is
an `intent(out)` argument that, if present, returns an error code
indicating any problem found in processing `init`, and if absent and
an error was found result in stopping processing with an informative
stop code. It can have any of the following error codes:

* `success` - no problem found

* `alloc_fault` - `self` was of type `bitset_large` and memory
  allocation failed

* `array_size_invalid_error` - bits was present with either a negative
  value, or a value greater than 64 when `self` was of type
  `bitset_64`.

#### Example

```fortran
    program demo_init
        use stdlib_bitsets
        type(bitset_large) :: set0
        call set0 % init(166)
        if ( set0 % bits() == 166 ) &
            write(*,*) `SET0 has the proper size.'
        if ( set0 % none() ) write(*,*) 'SET0 is properly initialized.'
    end program demo_init
```

### `input` - reads a bitset from an unformatted file

#### Status

Experimental

#### Description

Reads a bitset from its binary representation in an unformatted

> file.

Reviewer: Is it not worthwhile to mention the type of access used for unformatted files?

Author: I had been assuming sequential access, but it doesn't matter as long as the read is in the right position in the file. To ensure that you probably want to use the same access in the read as was used in the write.

#### Syntax

`call self % [[bitset_type(class):input(bound)]] (unit [, status])`

#### Class

Subroutine

#### Arguments

`self`: shall be a scalar variable of class `bitset_64` or
`bitset_large`. It is an `intent(out)` argument.

`unit`: shall be a scalar default integer expression. It is an
`intent(in)` argument. Its value must be that of a logical unit
number for an open unformatted file with `READ` or `READWRITE`
access positioned at the start of a BITSET value written by a
`bitset_type` `output` subroutine by the same processor.

`status` (optional): shall be a scalar default integer variable. If
present its value shall be of one of the error codes defined in this

> module. IF absent and it would have had a value other than `success`

Reviewer: ```suggestion
module. If absent and it would have had a value other than `success`
```

processing will stop with an informative stop code. Allowed error code
values for this `status` are:

* `success` - no problem found

* `alloc_fault` - `self` was of type `bitset_large` and allocation of
  memory failed.

* `array_size_invalid_error` - if the number of bits read from `unit`
  is either negative or greater than 64, if class of `self` is
  `bitset_64`.

* `read_failure` - failure during a read statement

#### Example

```fortran
    program demo_input
        character(*), parameter :: &
            bits_0   = '000000000000000000000000000000000', &
            bits_1   = '000000000000000000000000000000001', &
            bits_33  = '100000000000000000000000000000000'
        integer :: unit
        type(bitset_64) :: set0, set1, set2, set3, set4, set5
        call set0 % from_string( bits_0 )
        call set1 % from_string( bits_1 )
        call set2 % from_string( bits_33 )
        open( newunit=unit, file='test.bin', status='replace', &
            form='unformatted', action='write' )

>         call set2 % output(unit)

Reviewer: Maybe I missed it, but would it be possible to use:
```fortran
 call set2 % output('test.bin')
```
and similarly for `input`?

        call set1 % output(unit)
        call set0 % output(unit)
        close( unit )
        open( newunit=unit, file='test.bin', status='old', &
            form='unformatted', action='read' )
        call set5 % input(unit)
        call set4 % input(unit)
        call set3 % input(unit)
        close( unit )
        if ( set3 /= set0 .or. set4 /= set1 .or. set5 /= set2 ) then
            error stop 'Transfer to and from units using ' // &
                ' output and input failed.'
        else
            write(*,*) 'Transfer to and from units using ' // &
                'output and input succeeded.'
        end if
    end program demo_input
```

### `none` - determines whether no bits are set

#### Status

Experimental

#### Description

Determines whether no bits are set in `self`.

#### Syntax

`result = self % [[bitset_type(class):none(bound)]] ()`

#### Class

Elemental function.

#### Argument

`self`: shall be a scalar expression of class `bitset_type`. It is an
  `intent(in)` argument.

> #### Result character
> 
> The result is a default logical scalar.
> 
> #### Result value
> 
> The result is `.true.` if no bits in `self` are set, otherwise it is
> `.false.`.

Reviewer: ```suggestion
#### Result value

The result is a default logical scalar.
The result is `.true.` if no bits in `self` are set, otherwise it is
`.false.`.
```

#### Example

```fortran
    program demo_none
        use stdlib_bitsets
        character(*), parameter :: &
            bits_0 = '0000000000000000000'
        type(bitset_large) :: set0
        call set0 % from_string( bits_0 )
        if ( set0 % none() ) then
            write(*,*) "FROM_STRING interpreted " // &
                "BITS_0's value properly."
        end if
        call set0 % set(5)
        if ( .not. set0 % none() ) then
            write(*,*) "NONE interpreted SET0's value properly."
        end if
    end program demo_none
```

### `not` - Performs the logical complement on a bitset

#### Status

Experimental

#### Description

Performs the logical complement on the bits of `self`.

#### Syntax

> `result = self % [[bitset_type(class):not(bound)]] ()`

Reviewer: ```suggestion
`call self % [[bitset_type(class):not(bound)]] ()`
```

#### Class

Elemental subroutine.

#### Argument

`self` shall be a scalar variable of class `bitset_type`. It is an
  `intent(inout)` argument. On return its bits shall be the logical
  complement of their values on input.

#### Example

```fortran
    program demo_not
        use stdlib_bitsets
        type(bitset_large) :: set0
        call set0 % init( 155 )
        if ( set0 % none() ) then
            write(*,*) "FROM_STRING interpreted " // &
                "BITS_0's value properly."
        end if
        call set0 % not()
        if ( set0 % all() ) then
            write(*,*) "ALL interpreted SET0's value properly."
        end if
    end program demo_not
```

### `or` - Bitwise OR of the bits of two bitsets.

#### Status

Experimental

#### Description

Replaces the original bits of `set1` with the bitwise `or` of those
bits with the bits of `set2`. Note `set1` and `set2` must have the

> samee number of bits, otherwise the result is undefined.

Reviewer: ```suggestion
same number of bits, otherwise the result is undefined.
```

> samee number of bits, otherwise the result is undefined.

Reviewer: How is `undefined` defined? `NaN? (I didn't look to the code yet, but I think it would be good to define "undefined" in the specs.)

#### Syntax

`call [[stdlib_bitsets(module):or(interface)]](set1, set2)`

#### Class

Elemental subroutine.

#### Arguments

`set1`: shall be a scalar `bitset_64` or `bitset_large` variable. It
 is an `intent(inout)` argument. On return the values of the bits in
 `setf` are the bitwise `or` of the original bits in `set1` with the
 corresponding bits in `set2`.

`set2`: shall be a scalar expression of the same type as `set1`. It is

> an `intent(in)` argument. Note `bits(set2)` must equal `bits(set1)`

Reviewer: ```suggestion
an `intent(in)` argument. Note that `bits(set2)` must equal `bits(set1)`
```

otherwise the results are undefined.

#### Example

```fortran
    program demo_or
        use stdlib_bitsets
        type(bitset_large) :: set0, set1
        call set0 % init(166)
        call set1 % init(166)
        call or( set0, set1 ) ! none none
        if ( none(set0) ) write(*,*) 'First test of OR worked.'
        call set0 % not()
        call or( set0, set1 ) ! all none
        if ( all(set0) ) write(*,*) 'Second test of OR worked.'
        call set0 % not()
        call set1 % not()
        call or( set0, set1 ) ! none all
        if ( all(set0) ) write(*,*) 'Third test of OR worked.'
        call set0 % not()
        call or( set0, set1 ) ! all all
        if ( all(set0) ) write(*,*) 'Fourth test of OR worked.'
    end program demo_or
```

### `output` - Writes a binary representation of a bitset to a file

#### Status

Experimental

#### Description

> Writes a binary representation of a bitset to an unformatted file.

Reviewer: What kind of access is used? It might be important to mention it because the user must open the file himself.

`bits_kind`. It is an `intent(in)` argument.

`end_pos` (optional): shall be a scalar integer expression of kind
`bits_kind`. It is an `intent(in)` argument.

#### Example

```fortran
    program demo_set
        use stdlib_bitsets
        type(bitset_large) :: set0
        call set0 % init(166)
        if ( set0 % none() ) write(*,*) 'SET0 is properly initialized.'
        call set0 % set(165)
        if ( set0 % test(165) ) write(*,*) 'Bit 165 is set.'
        call set0 % set(0,164)
        if ( set0 % all() ) write(*,*) 'All bits are set.'
    end program demo_set
```

### `test` - determine whether a bit is set

#### Status

Experimental

#### Descriptions

Determine whether the bit at position `pos` is set to 1 in `self`.

#### Syntax

`result = self % [[bitset_type(class):test(bound)]](pos)`

#### Class

Elemental function.

#### Arguments

`self`: shall be a scalar expression of class `bitset_type`. It is an
`intent(in)` argument.

`pos`: shall be a scalar integer expression of kind `bits_kind`. It is
an `intent(in)` argument.

#### Result character

The result is a default logical scalar.

#### Result value

> The result is `.true.` if the bit at `pos`, in `self` is set,

Reviewer: ```suggestion
The result is `.true.` if the bit at `pos` in `self` is set,
```

otherwise it is `.false.`. If `pos` is outside the range
`0... bits(self)-1` the result is `.false.`.

> 

Reviewer: " If `pos` is outside the range ... `.false.`.": does this make sense? As a user, I would prefer that the program stops, or at least, returns something corresponding to `NaN`

#### Example

```fortran
    program demo_test
        use stdlib_bitsets
        type(bitset_large) :: set0
        call set0 % init(166)
        call set0 % not()
        if ( set0 % all() ) write(*,*) 'SET0 is properly initialized.'
        call set0 % clear(165)
        if ( .not. set0 % test(165) ) write(*,*) 'Bit 165 is cleared.'
        call set0 % set(165)
        if ( set0 % test(165) ) write(*,*) 'Bit 165 is set.'
    end program demo_test
```

### `to_string` - represent a bitset as a binary literal

### Status

Experimental

#### Description

Represents the value of `self` as a binary literal in `string`.

#### Syntax

> `call self % [[bitset_type(class):to_string(bound)]](string[, status])

Reviewer: ```suggestion
`call self % [[bitset_type(class):to_string(bound)]](string[, status])`
```

length. It is an `intent(out)` argument. On return it shall hav a
*binary-literal* representation of the bitset `self`.

`status` (optional): shall be a scalar default integer variable. It is
an `intent(out)` argument. If present it shall have either the value
`success` or `alloc_fault`. If absent and it would have had the value
`alloc_fault` then processing will stop with an informative test as
the stop code. The values have the following meanings:

`success` - no problem found.

`alloc_fault` - allocation of `string` failed.

#### Example

```fortran
    program demo_to_string
        use stdlib_bitsets
        character(*), parameter :: &
            bits_all = '111111111111111111111111111111111'
        type(bitset_64) :: set0
        character(:), allocatable :: new_string
        call set0 % init(33)
        call set0 % not()
        call set0 % to_string( new_string )
        if ( new_string == bits_all ) then
            write(*,*) "TO_STRING transferred BITS0 properly" // &
                " into NEW_STRING."
        end if
    end program demo_to_string
```

### `value` - determine the value of a bit

#### Status

> Experimeental

Reviewer: ```suggestion
Experimental
```

#### Description

Determines the value of the bit at position, `pos`, in `self`.

#### Syntax

`result = self % [[bitset_type(class):value(bound)]](pos)`

#### Class

Elemental function.

#### Arguments

`self`: shall be a scalar expression of class `bitset_type`. It is an
`intent(in)` argument.

`pos`: shall be a scalar integer expression of kind `bits_kind`. It is
an `intent(in)` argument.

> #### Result character
> 
> The result is a default logical scalar.
> 
> #### Result value
> 
> The result is one if the bit at `pos` in `self` is set, otherwise it
> is zero. If `pos` is outside the range `0... bits(set)-1` the result
> is zero.
> 

Reviewer: ```suggestion
#### Result value

The result is a default logical scalar.
The result is 1 if the bit at `pos` in `self` is set, otherwise it
is 0. If `pos` is outside the range `0... bits(set)-1` the result
is 0.

```

Again, does it make sense to return 0 if pos is outside the range?

  an `intent(in)` argument. Note `set1` and `set2` must have the
samee number of bits, otherwise the result is undefined.

#### Example

```fortran
    program demo_xor
        use stdlib_bitsets
        type(bitset_large) :: set0, set1
        call set0 % init(166)
        call set1 % init(166)
        call xor( set0, set1 ) ! none none
        if ( none(set0) ) write(*,*) 'First test of XOR worked.'
        call set0 % not()
        call xor( set0, set1 ) ! all none
        if ( all(set0) ) write(*,*) 'Second test of XOR worked.'
        call set0 % not()
        call set1 % not()
        call xor( set0, set1 ) ! none all
        if ( all(set0) ) write(*,*) 'Third test of XOR worked.'
        call set0 % not()
        call xor( set0, set1 ) ! all all
        if ( none(set0) ) write(*,*) 'Fourth test of XOR worked.'
    end program demo_xor
```

## Specification of the `stdlib_bitsets` operators

### `==` - compare two bitsets to determine whether the bits have the same value

#### Status

Experimental

#### Description

Returns `.true.` if all bits in `set1` and `set2` have the same value,
`.false.`  otherwise.

#### Syntax

> `Result = set1 [[stdlib_bitsets(module):==(interface)]] set2

Reviewer: ```suggestion
`result = set1 [[stdlib_bitsets(module):==(interface)]] set2
```

#### Class

Elemental operator

#### Arguments

`set1`: shall be a scalar `bitset_64` or `bitset_large` expression. It
is an `intent(in)` argument.

`set2`: shall be a scalar expression of the same type as `self`. It
will have the same number of bits as `set1`. It is an `intent(in)`
argument.

> #### Result character
> 
> The result is a default logical scalar.
> 
> #### Result value
> 
> The result is `.true.` if the bits in both bitsets are set
>   to the same value, otherwise the result is `.FALSE.`.

Reviewer: ```suggestion
#### Result value

The result is a default logical scalar.
The result is `.true.` if the bits in both bitsets are set
  to the same value, otherwise the result is `.false.`.
```

            set2 /= set2 ) then
            write(*,*) 'Passed 64 bit inequality tests.'
        else
            error stop 'Failed 64 bit inequality tests.'
        end if
    end program demo_inequality
```

### `>=` - compare two bitsets to determine whether the first is greater than or equal to the second

#### Status

Experimental

#### Description

Returns `.true.` if the bits in `set1` and `set2` are the same or the
highest order different bit is set to 1 in `set1` and to 0 in `set2`,
`.false.`.  otherwise. The sets must be the same size otherwise the
results are undefined

#### Syntax

`Result = set1 [[stdlib_bitsets(module):>=(interface)]] set2`

#### Class

Elemental operator

#### Arguments

`set1`: shall be a scalar `bitset_64` or `bitset_large` expression. It
is an `intent(in)` argument.

`set2`: shall be a scalar expression of the same type as `self`. It
will have the same number of bits as `set1`. It is an `intent(in)`
argument.

#### Result character

The result is a default logical scalar.

#### Result value

The result is `.true.` if the bits in `set1` and `set2` are the same
or the highest order different bit is set to 1 in `set1` and to 0 in

> `set2`, `.false.`.  otherwise.

Reviewer: ```suggestion
`set2`, `.false.`  otherwise.
```

#### Example

```fortran
    program demo_ge
        use stdlib_bitsets
        type(bitset_64) :: set0, set1, set2
        call set0 % init( 33 )
        call set1 % init( 33 )
        call set2 % init( 33 )
        call set1 % set( 0 )
        call set2 % set( 32 )
        if ( set1 >= set0 .and. set2 >= set1 .and. set2 >= set0 .and. &
            set0 >= set0 .and. set1 >= set1 .and. set2 >= set2 .and. &
            .not. set0 >= set1 .and. .not. set0 >= set2 .and. .not.   &
            set1 >= set2 ) then
            write(*,*) 'Passed 64 bit greater than or equals tests.'
        else
            error stop 'Failed 64 bit greater than or equals tests.'
        end if
    end program demo_ge
```

### `>` - compare two bitsets to determine whether the first is greater than the other

#### Status

Experimental

#### Description

Returns `.true.` if the bits in `set1` and `set2` differ and the
highest order different bit is set to 1 in `set1` and to 0 in `set2`,

> `.false.`.  otherwise. The sets must be the same size otherwise the

Reviewer: ```suggestion
`.false.`  otherwise. The sets must be the same size otherwise the
```

results are undefined

#### Syntax

> `Result = set1 [[stdlib_bitsets(module):>(interface)]] set2`

Reviewer: ```suggestion
`result = set1 [[stdlib_bitsets(module):>(interface)]] set2`
```

argument.

#### Result character

The result is a default logical scalar.

#### Result value

The result is `.true.` if the bits in `set1` and `set2` differ and the
highest order different bit is set to 1 in `set1` and to 0 in `set2`,
`.false.`.  otherwise.

#### Example

```fortran
    program demo_gt
        use stdlib_bitsets
        type(bitset_64) :: set0, set1, set2
        call set0 % init( 33 )
        call set1 % init( 33 )
        call set2 % init( 33 )
        call set1 % set( 0 )
        call set2 % set( 32 )
        if ( set1 > set0 .and. set2 > set1 .and. set2 > set0 .and. &
            .not. set0 > set0 .and. .not. set0 > set1 .and. .not.   &
            set1 > set2 ) then
            write(*,*) 'Passed 64 bit greater than tests.'
        else
            error stop 'Failed 64 bit greater than tests.'
        end if
    end program demo_gt
```

### `<=` - compare two bitsets to determine whether the first is less than or equal to the other

#### Status

Experimental

#### Description

Returns `.true.` if the bits in `set1` and `set2` are the same or the
highest order different bit is set to 0 in `set1` and to 1 in `set2`,

> `.false.`.  otherwise. The sets must be the same size otherwise the

Reviewer: ```suggestion
`.false.`  otherwise. The sets must be the same size otherwise the
```

results are undefined

#### Syntax

> `Result = set1 [[stdlib_bitsets(module):<=(interface)]] set2`

Reviewer: ```suggestion
`result = set1 [[stdlib_bitsets(module):<=(interface)]] set2`
```

#### Class

Elemental operator

#### Arguments

`set1`: shall be a scalar `bitset_64` or `bitset_large` expression. It
is an `intent(in)` argument.

`set2`: shall be a scalar expression of the same type as `self`. It
will have the same number of bits as `set1`. It is an `intent(in)`
argument.

#### Result character

The result is a default logical scalar.

#### Result value

The result is `.true.` if the bits in `set1` and `set2` are the same
or the highest order different bit is set to 0 in `set1` and to 1 in

> `set2`, `.false.`.  otherwise.

Reviewer: ```suggestion
`set2`, `.false.`  otherwise.
```

#### Example

```fortran
    program demo_le
        use stdlib_bitsets
        type(bitset_64) :: set0, set1, set2
        call set0 % init( 33 )
        call set1 % init( 33 )
        call set2 % init( 33 )
        call set1 % set( 0 )
        call set2 % set( 32 )
        if ( set0 <= set1 .and. set1 <= set2 .and. set0 <= set2 .and. &
            set0 <= set0 .and. set1 <= set1 .and. set2 <= set2 .and. &
            .not. set1 <= set0 .and. .not. set2 <= set0 .and. .not.   &
            set2 <= set1 ) then
            write(*,*) 'Passed 64 bit less than or equal tests.'
        else
            error stop 'Failed 64 bit less than or equal tests.'
        end if
    end program demo_le
```

### `<` - compare two bitsets to determine whether the first is less than the other

#### Status

Experimental

#### Description

Returns `.true.` if the bits in `set1` and `set2` differ and the
highest order different bit is set to 0 in `set1` and to 1 in `set2`,

> `.false.`.  otherwise. The sets must be the same size otherwise the

Reviewer: ```suggestion
`.false.`  otherwise. The sets must be the same size otherwise the
```

results are undefined

#### Syntax

> `Result = set1 [[stdlib_bitsets(module):<(interface)]] set2`

Reviewer: ```suggestion
`result = set1 [[stdlib_bitsets(module):<(interface)]] set2`
```
